src/generate/generic.rs

```diff
@@ -18,23 +30,97 @@ pub fn render() -> Result<Vec<Tokens>> {
     });
 
     generic_items.push(quote! {
-        ///Value read from the register
-        pub struct FR<U, T> {
-            pub(crate) bits: U,
-            _reg: marker::PhantomData<T>,
+        ///Wrapper for registers
+        pub struct Reg<U, REG> {
+            register: vcell::VolatileCell<U>,
+            _marker: marker::PhantomData<REG>,
         }
 
-        impl<U, T, FI> PartialEq<FI> for FR<U, T>
+        impl<U, REG> Reg<U, REG>
         where
-            U: PartialEq,
-            FI: ToBits<U>
+            Self: Readable,
+            U: Copy
         {
-            fn eq(&self, other: &FI) -> bool {
-                self.bits.eq(&other._bits())
+            ///Reads the contents of the register
+            #[inline(always)]
+            pub fn read(&self) -> R<U, Self> {
+                R {bits: self.register.get(), _reg: marker::PhantomData}
+            }
+        }
+
+        impl<U, REG> Reg<U, REG>
+        where
+            Self: ResetValue<U> + Writable,
+            U: Copy,
+        {
+            ///Writes the reset value to the register
+            #[inline(always)]
+            pub fn reset(&self) {
+                self.register.set(Self::reset_value())
             }
         }
+    });
 
-        impl<U, T> FR<U, T>
+    generic_items.push(quote! {
+        impl<U, REG> Reg<U, REG>
+        where
+            Self: ResetValue<U> + Writable,
+            U: Copy
+        {
+            ///Writes to the register
+            #[inline(always)]
+            pub fn write<F>(&self, f: F)
+            where
+                F: FnOnce(&mut W<U, Self>) -> &mut W<U, Self>
+            {
+                self.register.set(f(&mut W {bits: Self::reset_value(), _reg: marker::PhantomData}).bits);
+            }
+        }
+    });
+
+    generic_items.push(quote! {
+        impl<U, REG> Reg<U, REG>
+        where
+            Self: Writable,
+            U: Copy + Default
```

Comment:
Is this `Default` equals to "returns zero"?

Reply:
Yes, it is for `bool`, `u8`, `u16` `u32` and `u64`.

Reply:
Yes, but these types are not enforced I guess

Reply:
They are enforced by size type of register or field.

Reply:
But it's possible to create a wrapper with the same size and a custom Default impl that returns reset value, for example

Reply:
How? All fields are private. You can get register only from peripheral as it was before.

Reply:
Hmm, now I see. Thanks.